src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1125 @@
+
+            EXAMPLES::
+
+                sage: F = IntegerValuedPolynomialRing(GF(7)).S(); F
+                Integer-Valued Polynomial Ring over Finite Field of size 7
+                in the shifted basis
+
+            Elements of the integer-valued polynomial ring canonically
+            coerce in::
+
+                sage: x = F.gen()
+                sage: F.coerce(x*x) # indirect doctest
+                6*S[1] + 2*S[2]
+
+            Elements of the integers coerce in, since there is a coerce map
+            from `\ZZ` to GF(7)::
+
+                sage: F.coerce(1)       # indirect doctest
+                S[0]
+
+            There is no coerce map from `\QQ` to `\GF{7}`::
+
+                sage: F.coerce(2/3)  # indirect doctest
+                Traceback (most recent call last):
+                ...
+                TypeError: no canonical coercion from Rational Field to
+                Integer-Valued Polynomial Ring over Finite Field of size 7
+                in the shifted basis
+
+            Elements of the base ring coerce in::
+
+                sage: F.coerce(GF(7)(5))
+                5*S[0]
+
+            The integer-valued polynomial ring over `\ZZ` on `x` coerces in,
+            since `\ZZ` coerces to `\GF{7}`::
+
+                sage: G = IntegerValuedPolynomialRing(ZZ).S()
+                sage: Gx = G.gen()
+                sage: z = F.coerce(Gx**2); z
+                -S[1] + 2*S[2]
+                sage: z.parent() is F
+                True
+
+            However, `\GF{7}` does not coerce to `\ZZ`, so the shuffle
+            algebra over `\GF{7}` does not coerce to the one over `\ZZ`::
+
+                sage: G.coerce(x^3+x)
+                Traceback (most recent call last):
+                ...
+                TypeError: no canonical coercion from Integer-Valued Polynomial
+                Ring over Finite Field of size 7  in the shifted basis
+                to Integer-Valued Polynomial
+                Ring over Integer Ring in the shifted basis
+
+            TESTS::
+
+                sage: F = IntegerValuedPolynomialRing(ZZ).S()
+                sage: G = IntegerValuedPolynomialRing(QQ).S()
+                sage: H = IntegerValuedPolynomialRing(ZZ).S()
+                sage: F._coerce_map_from_(G)
+                False
+                sage: G._coerce_map_from_(F)
+                True
+                sage: F._coerce_map_from_(H)
+                True
+                sage: F._coerce_map_from_(QQ)
+                False
+                sage: G._coerce_map_from_(QQ)
+                True
+                sage: F.has_coerce_map_from(PolynomialRing(ZZ,'x'))
+                False
+            """
+            # integer-valued polynomial rings over any base
+            # that coerces in:
+            if isinstance(R, IntegerValuedPolynomialRing.Shifted):
+                return self.base_ring().has_coerce_map_from(R.base_ring())
+            return self.base_ring().has_coerce_map_from(R)
```

Comment:
This would also be a good place to do the coercion from the other basis (or bases should more come along).

Reply:
better than the pair of `register_as_coercion` ?

Reply:
I think so. Both bases do not need to be created an nailed in memory (at creation of the abstract object), and it has one (standard) centralized place for setting the coercions. (It also means we don't have to worry about the initialization code creating the coercion maps `B -> S`  and `S -> B` becoming defunct because one of `B` or `S` gets garbage collected and then recreated, which then leads to further issues. This is something that has happened in the past.)